### PR TITLE
Added threading document and initial draft of relates_to.

### PR DESCRIPTION
**Pull request by Half-Shot**
Basic draft of how I see it fitting together. Feedback?

### CODE REVIEW

File: drafts/general_api.rst [hunk 1]
```diff
@@ -1083,65 +1083,28 @@ needs to know how to combine the events.
...
-clients may receive the events separately down the event stream.
-
-TODO:
- - Can a child event reply to multiple parent events? Use case?
- - Should a parent event and its children share a thread ID? Does the 
-   originating HS set this ID? Is this thread ID exposed through federation? 
-   e.g. can a HS retrieve all events for a given thread ID from another HS?
 
-   
+See ``threading.rst``
```
Comment (Kegsay) [inline drafts/general_api.rst]: Not sure how I feel about splitting this out from the general API and having to play "hunt the file" to get at the information. We can do this linking once we feel we have the right solution for threading.

File: drafts/event_threading.rst [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+- Replies are made as follows:
+
+::
+
+ ┌B─D─E─I─J─K
+ A
+ ├C─F
+ │└─G
+ └H
+ └L
```
Comment (Kegsay) [inline drafts/event_threading.rst]: This is hard to follow. I image there is subtle nesting going on here but it's not clear. This is further confused by the assertion "A has two replies, B and C" whereas I think H is also a reply to A?

File: drafts/event_threading.rst [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+Child events should be bundled with the parent event under the ``relates_to``
+key. There should also be two limits to how many events are returned:
+
+- A limit on the amount of events returned per thread. ``depth_limit``
+- A limit on the amount of threads returned per events. ``thread_limit``
+
+The ``relates_to`` key on an event will contain an array of every relation to
+that event. The relations will contain a ``start`` and ``next`` key for
+pagination. The start key is important should the client not recieve previous
+events since it will be the only way to determine the root event. The chain of
```
Comment (Kegsay) [inline drafts/event_threading.rst]: > The start key is important should the client not recieve previous events since it will be the only way to determine the root event.

Wording feels wrong, and typo `recieve`. I also debate whether the start key is important enough to be highlighted as such.

> The start key can be used to determine the root event of the thread. Alternatively, clients can paginate back until they reach the root event.

Perhaps?

File: drafts/event_threading.rst [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+events in this thread (where each event is a response to the previous event)
```
Comment (Kegsay) [inline drafts/event_threading.rst]: Strictly 1 event in response to 1 previous event? Or can an event have 2 previous events to merge threads? Mention either way.

File: drafts/event_threading.rst [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    relates_to:[
+      {
+        "start":"ROOTTOKEN",
+        "events":[],
+        "next":null
+      }
+    ]
+  }
+
+In this example, F is isolated and must use start to find the root event.
```
Comment (Kegsay) [inline drafts/event_threading.rst]: Backtick "start" please :)

File: drafts/event_threading.rst [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      },
+      {
+        "start":"ROOTTOKEN",
+        "events":null,
+        "next":"TOKEN3"
+      }
+    ]
+  }
+  
+An example where a client recieves Message F but is missing some of the timeline.
```
Comment (Kegsay) [inline drafts/event_threading.rst]: Typo: `recieves`

File: drafts/general_api.rst (hunk 1 above)
Comment (Half-Shot) [inline drafts/general_api.rst]: My reasoning was that this is quite a substantial change and that it was getting starting to take up quite a lot of room in general_api. However, it does make sense not to split it out until it's considered ready.

**Comment by Kegsay** [comment]
There's a decent amount of cruft from the general API which you've brought over to the new threading file. Please leave it in the general API and stick strictly to the proposed threading solution in the threading file, else it makes it harder to follow.

It would be worthwhile if you were consistent with your terminology as well. There's a lot of very similar moving parts in threading (multiple `limits`, multiple pagination tokens, different kinds of relations between events, etc) so be careful with your wording: `replies` vs `children`, `message` vs `event`, the use of unqualified "chain" (of child events? Of children and parents? Of just parents?), the use of unqualified "limits" (e.g. "H and L are two more responses that were not included to demonstrate limits."), etc.

In these kinds of problems, visual explanations through lots of ASCII diagrams tend to convey the solution better than wordy words.

A good start overall though :D it would be good to see how this proposed solution addresses the 3 main use cases we have for threading currently:
 - Reddit style threads (**all** the nesting)
 - No deep nested threads (1 level deep)
 - Vector-style flinging

along with any limitations on your design:
 - Number of round trips
 - Amount of unwanted data sent to clients
 - Ease of extraction from the JSON structure

File: drafts/general_api.rst (hunk 1 above)
Comment (Kegsay) [inline drafts/general_api.rst]: I'm happy that you've made a new file for it, I'm just not happy you removed the existing stuff from general API :)

**Comment by ara4n** [comment]
it may be easier to draft this with feedback in google docs, and then port it into formal RST once everything has shaken out?  (This is the structure we're trying in the core team currently)

**Comment by Half-Shot** [comment]
@Kegsay Thanks for the feedback. I'll make the changes and put them in the doc.

@ara4n Sounds like a plan. I'll get it set up and post it in core?

**Comment by richvdh** [comment]
Let's close this until @Half-Shot has got a bit closer to having something ready to ship
